src/DurableTask.Netherite/StorageLayer/Faster/StoreWorker.cs

```diff
@@ -349,7 +371,126 @@ void ScheduleNextIdleCheckpointTime()
             var actual = (((earliest - offset - 1) / period) + 1) * period + offset;
             this.timeOfNextIdleCheckpoint = new DateTime(actual, DateTimeKind.Utc);
         }
-        
+
+        void StartCheckpointOrFailOnTimeout(Func<bool> checkpointRoutine, string messageOnError)
+        {
+            var checkpointStarted = checkpointRoutine.Invoke();
+            if (checkpointStarted)
+            {
+                this.timeOfFirstRefusedCheckpoint = null; 
+            }
+            else
+            {
+                // track start of FASTER refusal to start checkpoint
+                var currentTime = DateTime.UtcNow;
+                this.timeOfFirstRefusedCheckpoint ??= currentTime;
+
+                // if the refusal to checkpoint started over a minute ago, terminate partition
+                TimeSpan duration = currentTime - this.timeOfFirstRefusedCheckpoint.Value;
+                if (duration > TimeSpan.FromMinutes(1))
+                {
+                    messageOnError += $" FASTER first refused to checkpoint at '{this.timeOfFirstRefusedCheckpoint}'. Duration of refusal = {duration}";
```

Comment:
nit: string is starting with an extra space
I dont' think you need to add quotes around a time.

Reply:
This is actually on purpose. Since we're adding new text to the error, I need to separate it with a space. I'm adding a dot (".") in there to make this clearer.